### PR TITLE
Addition of the class of the compound lens model and functions to generate the delector populations

### PR DESCRIPTION
**Pull request by kta-cosmo**
# Pull Request - Compound lens model 

## Overview
We have added codes for the deflector population from the compound lens model in slsim. We have added the funtions for both reading files generated from our code and for generating in slsim.

## Changes Made
Detail the major changes in this pull request.
- Added `compound_lens_halos_galaxies.py` in `slsim/slsim/Deflectors`
  - **Purpose**: Implements a class named `CompoundLensHalosGalaxies`.
- Added `sl_hammocks_pipeline.py` in `slsim/slsim/Pipelines`
  - **Purpose**: Implements a class named `SLHammocksPipeline` and imports `colossus.cosmology` package.
- Added `halo_population.py` in `slsim/slsim/Pipelines`
  - **Purpose**: Function to generate the deflector population for dark matter halos.
- Added `galaxy_population.py` in `slsim/slsim/Pipelines`
  - **Purpose**: Function to generate the deflector population for galaxies; imports `colossus.halo`.
- Added `pop_salp_a1_zl001_to_5_wo_sub.csv` in `slsim/data/SL-Hammocks`
  - **Purpose**: CSV file containing deflector population data generated from the SL-Hammocks code.
- Modified `lens_pop.py`
  - **Changes**: Added an input parameter named `slhammocks_config`. Added a new deflector type, "halo-models".

## Testing Done

Describe how the new and modified code has been tested.
- Tests for `CompoundLensHalosGalaxies` class
- Tests for `SLHammocksPipeline` functionality.
- Verification of the deflector population generation for both halos and galaxies.
By executing the following code:

```Python
from astropy.cosmology import FlatLambdaCDM
from astropy.units import Quantity
from slsim.lens_pop import LensPop

cosmo = FlatLambdaCDM(H0=70, Om0=0.3)
sky_area = Quantity(value=0.001, unit="deg2")
kwargs_deflector_cut ={"z_min": 0.01, "z_max": 2.0}
kwargs_source_cut = {"band": "g", "band_max": 28, "z_min": 0.1, "z_max": 5.0}

skypy_config='/your-path-to-slsim/data/SkyPy/lsst-like.yml'
slhammocks_config='/your-path-to-slsim/data/SL-Hammocks/pop_salp_a0001_zl001_to_5_wo_sub.csv'
gg_lens_pop = LensPop(
    deflector_type="halo-models",
    source_type="galaxies",
    kwargs_deflector_cut=kwargs_deflector_cut,
    kwargs_source_cut=kwargs_source_cut,
    kwargs_mass2light=None,
    skypy_config=skypy_config,
    slhammocks_config=slhammocks_config,
    sky_area=sky_area,
    cosmo=cosmo,
)
```
and
```Python
from astropy.cosmology import FlatLambdaCDM
from astropy.units import Quantity
from slsim.lens_pop import LensPop

cosmo = FlatLambdaCDM(H0=70, Om0=0.3)
sky_area = Quantity(value=0.001, unit="deg2")
kwargs_deflector_cut ={"z_min": 0.01, "z_max": 2.0}
kwargs_source_cut = {"band": "g", "band_max": 28, "z_min": 0.1, "z_max": 5.0}

skypy_config='/your-path-to-slsim/data/SkyPy/lsst-like.yml'
slhammocks_config=None
gg_lens_pop = LensPop(
    deflector_type="halo-models",
    source_type="galaxies",
    kwargs_deflector_cut=kwargs_deflector_cut,
    kwargs_source_cut=kwargs_source_cut,
    kwargs_mass2light=None,
    skypy_config=skypy_config,
    slhammocks_config=slhammocks_config,
    sky_area=sky_area,
    cosmo=cosmo,
)
```

## Additional Notes
This codes require to install [colossus]("https://bdiemer.bitbucket.io/colossus/") packages.

### CODE REVIEW

File: .github/pull_request_template.md
Comment (sibirrer) [inline .github/pull_request_template.md:1]: it looks like you modified/added not just the pull request template, but also the description of your specific pull request

File: data/SL-Hammocks/pop_salp_a0001_zl001_to_5_wo_sub.csv
Comment (sibirrer) [inline data/SL-Hammocks/pop_salp_a0001_zl001_to_5_wo_sub.csv:1]: can you add a short readme of what this csv file is and what the columns mean? Otherwise we might risk accidentally removing it or changing it in undesired ways

File: slsim/Deflectors/DeflectorTypes/nfw_hernquist.py
Comment (sibirrer) [inline slsim/Deflectors/DeflectorTypes/nfw_hernquist.py:29]: this line caused an error in the CI. Here you require the _deflector_dict object to be an astropy table, while in the input the format is a dictionary (compatible with astropy table, but not vise versa)

File: slsim/Deflectors/DeflectorTypes/nfw_hernquist.py
Comment (sibirrer) [inline slsim/Deflectors/DeflectorTypes/nfw_hernquist.py:51]: perhaps add gel_disp to the deflector_dict after calculations happened. Then we would not need to re-compute it if we want to access this number multiple times for the same deflector (just a thought I had and should have done myself before)

**Review by sibirrer — CHANGES_REQUESTED** [review]
thank you very much @kta-cosmo! I have a few higher level requests, since the continues integration fails (see the tests). Then the formatting is also failing (check the error message on this thread and let me know if there are issues). I know these are rather strict but hopefully they will work.
I will then go into a bit more detail and will also ask you to write unit tests in the /tests/ folder mirroring the package. You can also execute those locally with py.test /tests/<folder or file path to be tested>.
Thank you very much!

File: slsim/Pipelines/sl_hammocks_pipeline.py
Comment (sibirrer) [inline slsim/Pipelines/sl_hammocks_pipeline.py:100]: can you add document strings for this function?

File: slsim/Pipelines/halo_population.py
Comment (sibirrer) [inline slsim/Pipelines/halo_population.py:1]: can this file be moved to the Deflector folder? I think it's better suited there as it deals with the halo population

**Review by sibirrer — CHANGES_REQUESTED** [review]
Thank you very much @kta-cosmo! I have some minor comments and suggestions.
Then, can you also add document strings for those definitions missing them and add test functions? When we see how the code is tested it is easier to review individual lines.
Thank you very much!

**Comment by ajshajib** [comment]
@kta-cosmo, please pull from the main branch to sync with the latest version of that. Then, the coverage report will start working for this PR.

File: slsim/Deflectors/compound_lens_halos_galaxies.py
Comment (sibirrer) [inline slsim/Deflectors/compound_lens_halos_galaxies.py:127]: I made quite a few changes of this function here: https://github.com/LSST-strong-lensing/slsim/blob/6cffdf68434f343ab46f77534bf21db1e0a934cc/slsim/Deflectors/elliptical_lens_galaxies.py#L105 
Perhaps you will be able to use this one instead? Note my previous mistake of cos(2phi) instead of cos(phi)

**Review by sibirrer — CHANGES_REQUESTED** [review]
Hi @kta-cosmo, I just noticed you took a definition that I recently updated (and fixed some parts in), and though I let you know. Please ping me when/if you want a next review (by now it's going to be mainly documentation and testing things in the supporting files - thank you very much!

File: slsim/Deflectors/compound_lens_halos_galaxies.py
Comment (kta-cosmo) [reply slsim/Deflectors/compound_lens_halos_galaxies.py]: I changed to use the elliptical_projected_eccentricity function in slsim/slsim/Deflectors/elliptical_lens_galaxies.py when calculating the eccentricity of galaxies in compound_lens_halos_galaxies.py.

File: notebooks/Test_Lens_class.ipynb
Comment (sibirrer) [inline notebooks/Test_Lens_class.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #33.        deflector_type="all-galaxies",
</pre><p>here there are twice the deflector_type set, which also causes and error in the pre-commit CI</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/LSST-strong-lensing/slsim/pull/153/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='76'/>

File: slsim/lens.py
Comment (sibirrer) [inline slsim/lens.py:300]: is there a way you want to check that or do you want us to check it on our side?

File: slsim/lens_pop.py
Comment (sibirrer) [inline slsim/lens_pop.py:144]: is there a way to test a simple LensPop() instance with SLHammock? Just to make sure the integration works

**Review by sibirrer — APPROVED** [review]
Looks really good. I have to very minor points. I approve it and with the suggestions hand it over to @nkhadka21 to have a look.

**Comment by sibirrer** [comment]
@kta-cosmo I approved this PR from my side. I have to minor suggestions and hand it over to @nkhadka21.
There might then be a few overlaps in the halo rendering with the PR from @trivialTZ but we can look at it once both are merged since you both worked in parallel on it (for different purposes)

File: slsim/lens_pop.py
Comment (nkhadka21) [inline slsim/lens_pop.py:150]: If someone choose deflector_type == "halo-models" and slhammocks_config = None, then I think you should put slhammocks_config = "slsim/data/SL-Hammocks/pop_salp_a0001_zl001_to_5_wo_sub.csv" else: slhammocks_config = slhammocks_config. Meaning we should provide a default path if data is already strored inside the code.

**Review by nkhadka21 — CHANGES_REQUESTED** [review]
@kta-cosmo , thank you very much for this PR and all the effort that you put into it! I have some comments. Please address them and ping me when you are ready for the next review.

File: slsim/lens_pop.py
Comment (kta-cosmo) [reply slsim/lens_pop.py:142]: For example, I think you can run the following commands to test a simple LensPop() instance with "halo-models."

from astropy.cosmology import FlatLambdaCDM
from astropy.units import Quantity
from slsim.lens_pop import LensPop

cosmo = FlatLambdaCDM(H0=70, Om0=0.3, Ob0 = 0.05)
sky_area = Quantity(value=0.001, unit="deg2")

kwargs_deflector_cut = None
kwargs_source_cut = {"band": "g", "band_max": 28, "z_min": 0.1, "z_max": 5.0}

skypy_config = "/path/to/data/SkyPy/lsst-like.yml"
slhammocks_config = "/path/to/data/SL-Hammocks/pop_salp_a0001_zl001_to_5_wo_sub.csv"
gg_lens_pop = LensPop(
    deflector_type="halo-models",
    source_type="galaxies",
    kwargs_deflector_cut=kwargs_deflector_cut,
    kwargs_source_cut=kwargs_source_cut,
    kwargs_mass2light=None,
    skypy_config=skypy_config,
    slhammocks_config=slhammocks_config,
    sky_area=sky_area,
    cosmo=cosmo,
)
gg_lens_pop._lens_galaxies.draw_deflector()

File: slsim/Pipelines/sl_hammocks_pipeline.py
Comment (kta-cosmo) [reply slsim/Pipelines/sl_hammocks_pipeline.py]: I'm sorry if I misunderstood something, but the input parameter ellipticity is 1-q, where q is the axis ratio, which is the ellipticity in our SL-Hammocks calculation.
This function is for converting 1-q to (1 - q^2)/(1 + q^2).

File: slsim/lens_pop.py
Comment (kta-cosmo) [reply slsim/lens_pop.py:148]: Now, if slhammocks_config == None, we are trying to create halo and galaxy populations not using the csv file. Does this mean that we should have a separate switch for generating halo and galaxy populations, separately from slhammocks_config?

Comment (nkhadka21) [reply slsim/lens_pop.py:148]: @kta-cosmo , thanks for the clarification. My comment was not relevant here. What you have done is correct. No need to change anything.

**Comment by nkhadka21** [comment]
Hi @kta-cosmo , we are almost to the copletion of this PR! Thank you very much for your effort. I have just replied to your questions. Once you make these minor changes. I will merge this PR!

**Comment by sibirrer** [comment]
@kta-cosmo the latest merge fixed the coolest routines and if you pull the changes from the origin/main then your tests should pass

File: slsim/lens.py
Comment (kta-cosmo) [reply slsim/lens.py:304]: As I asked in Slack before, it looks like I can calculate the Einstein radius properly for massive halos (+ galaxies), but for small masses, it returns nan. I'm not sure if this is a bad thing in other calculation parts.

Comment (sibirrer) [reply slsim/lens.py:304]: @kta-cosmo for small halos, there may not be a Einstein radius as the lensing convergence never exceeds unity even in the center. We could replace nan with zero in the code for these cases if needed

Comment (kta-cosmo) [reply slsim/lens.py]: Thank you. Yes, I think so. I'll leave it to you to decide if nan should be changed to zero. I removed the TODO comment.

**Review by nkhadka21 — APPROVED** [review]
@kta-cosmo , thank you very much for your contribution! I approve this PR and hope to have more contributions from you in the future.

**Pull request merged**
